utils.py

```diff
@@ -47,6 +49,71 @@ class File(TypedDict):
 LOG_PATH = "./logs/"
 
 
+class Meteo_client(object):
+    # code is courtesy of Météo France: https://portail-api.meteofrance.fr/web/fr/faq
+
+    def __init__(self):
+        self.session = requests.Session()
+
+    def request(self, method, url, **kwargs):
+        # First request will always need to obtain a token first
+        if 'Authorization' not in self.session.headers:
+            self.obtain_token()
+        # Optimistically attempt to dispatch request
+        response = self.session.request(method, url, **kwargs)
+        if self.token_has_expired(response):
+            # We got an 'Access token expired' response => refresh token
+            self.obtain_token()
+            # Re-dispatch the request that previously failed
+            response = self.session.request(method, url, **kwargs)
+        return response
+
+    def get(self, url, **kwargs):
+        return self.request("GET", url, **kwargs)
+
+    def token_has_expired(self, response):
+        status = response.status_code
+        content_type = response.headers['Content-Type']
+        if status == 401 and 'application/json' in content_type:
+            if 'Invalid JWT token' in response.json()['description']:
+                return True
+        return False
+
+    def obtain_token(self):
+        # if threads are synchronous (for example on start), this should space them out
+        time.sleep(random.randint(1, 5))
+        if "token.json" not in os.listdir():
+            # Obtain new token
+            access_token_response = requests.post(
+                "https://portail-api.meteofrance.fr/token",
```

Comment:
I would define this endpoint url for as a class param. It would be easier to change and keep track of.